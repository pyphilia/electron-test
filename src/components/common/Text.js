import React from 'react';
import PropTypes from 'prop-types';
import ReactQuill from 'react-quill';
import { hasMath, renderMath } from '../../utils/math';

const modules = {
  toolbar: false,
};

const formats = [
  'header',
  'bold',
  'italic',
  'underline',
  'strike',
  'blockquote',
  'list',
  'bullet',
  'indent',
  'link',
  'image',
  'formula',
];

export const Text = ({ content, style, className, id }) => {
  let parsedContent = content;
  if (hasMath(content)) {
    parsedContent = renderMath(parsedContent);
  }
  return (
    <div style={style} id={id}>
      <ReactQuill
        className={className}
        value={parsedContent}
        modules={modules}
        formats={formats}
        style={{ border: '0' }}
        readOnly
      />
    </div>
  );
};

Text.propTypes = {
  content: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.shape({}),
  id: PropTypes.string,
};

Text.defaultProps = {
  content: '',
  className: '',
  style: {},
  id: null,
};

export default Text;
