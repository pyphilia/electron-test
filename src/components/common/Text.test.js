import React from 'react';
import { shallow } from 'enzyme';
import ReactQuill from 'react-quill';
import { Text } from './Text';
import * as FIXTURES from '../../test/fixtures/text';

const createTextProps = content => ({
  content,
  style: { color: 'red' },
  className: 'textClass',
});

describe('<Text />', () => {
  describe('trivial cases', () => {
    let wrapper;
    let props;

    beforeAll(() => {
      props = createTextProps('content');
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Text {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <div /> with correct style', () => {
      const div = wrapper.find('div');
      expect(div).toHaveLength(1);

      expect(div.prop('style')).toEqual(props.style);
    });

    it('renders one <ReactQuill /> component with correct className and simple content', () => {
      const reactQuill = wrapper.find(ReactQuill);
      expect(reactQuill).toHaveLength(1);

      expect(reactQuill.prop('className')).toEqual(props.className);
      expect(reactQuill.prop('value')).toEqual(props.content);
    });
  });

  describe('with math', () => {
    const cases = [
      ['x^2', FIXTURES.input1, FIXTURES.output1],
      ['quadratic', FIXTURES.quadraticInput, FIXTURES.quadraticOutput],
      ['christoffel', FIXTURES.christoffelInput, FIXTURES.christoffelOutput],
    ];

    it.each(cases)(`renders %s correctly with math`, (name, input, output) => {
      const props = createTextProps(input);
      // eslint-disable-next-line react/jsx-props-no-spreading
      const wrapper = shallow(<Text {...props} />);
      const reactQuill = wrapper.find(ReactQuill);
      expect(reactQuill.prop('value')).toEqual(output);
    });
  });
});
