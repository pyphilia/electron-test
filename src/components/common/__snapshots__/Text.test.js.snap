// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> trivial cases renders correctly 1`] = `
<div
  id={null}
  style={
    Object {
      "color": "red",
    }
  }
>
  <Quill
    className="textClass"
    formats={
      Array [
        "header",
        "bold",
        "italic",
        "underline",
        "strike",
        "blockquote",
        "list",
        "bullet",
        "indent",
        "link",
        "image",
        "formula",
      ]
    }
    modules={
      Object {
        "toolbar": false,
      }
    }
    readOnly={true}
    style={
      Object {
        "border": "0",
      }
    }
    theme="snow"
    value="content"
  />
</div>
`;
